### PR TITLE
support shardingjdbc database storage feature 

### PR DESCRIPTION
**Pull request by chidaodezhongsheng**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] New feature provided

___
### New feature or improvement
  Support shardingjdbc database storage feature, the users can use database to storage data instead of elasticsearch.

### CODE REVIEW

**Comment by wu-sheng** [comment]
@chidaodezhongsheng Conflicts must be resolved.

File: apm-collector/apm-collector-boot/src/main/resources/application.yml [hunk 1]
```diff
@@ -1,84 +1,90 @@
...
+    indexShardsNumber: 2
+    indexReplicasNumber: 0
+    highPerformanceMode: true
+    ttl: 7
+#storage:
+#  h2:
+#    url: jdbc:h2:~/memorydb
+#    userName: sa
+#storage:
+#  shardingjdbc:
```
Comment (wu-sheng) [inline apm-collector/apm-collector-boot/src/main/resources/application.yml]: ShardingJDBC is an incubating feature, no use cases yet. And none of the contributors are committer yet. So, by that, please don't put Sharding JDBC in default application.yml. Even it is under comments.

File: apm-dist/release-docs/LICENSE [hunk 2]
```diff
@@ -255,6 +256,7 @@ The text of each license is also included at licenses/LICENSE-[project].txt.
     Apache: commons-configuration 1.8: https://github.com/apache/commons-configuration, Apache 2.0
     Apache: commons-io 2.4: https://github.com/apache/commons-io, Apache 2.0
     Apache: tomcat 8.5.27: https://github.com/apache/tomcat/tree/trunk, Apache 2.0
+    Apache: groovy 2.4.5-indy: https://github.com/apache/groovy, Apache 2.0
```
Comment (wu-sheng) [inline apm-dist/release-docs/LICENSE]: Groovy has a NOTICE file. Please change NOTICE in same folder. And follow the file structure. Ref here: https://github.com/apache/groovy/blob/master/NOTICE

File: docs/README.md [hunk 3]
```diff
@@ -1,45 +1,46 @@
...
+    * [Quick start](en/Quick-start.md)
+    * [Supported middlewares, frameworks and libraries](Supported-list.md)
+      * [How to disable plugins?](en/How-to-disable-plugin.md)
+  * Advanced Features
+    * [Override settings through System.properties or System.env](en/Setting-override.md)
+    * [Direct uplink and disable naming discovery](en/Direct-uplink.md)
+    * [Open TLS](en/TLS.md)
+    * [Namespace Isolation](en/Namespace.md)
+    * [Token Authentication](en/Token-auth.md)
+    * [Open Database Sharding Storage](en/Shardingjdbc.md)
```
Comment (wu-sheng) [inline docs/README.md]: This should a section in `Incubating Features`. named: `Use Sharding JDBC as storage implementor`.

File: docs/en/Shardingjdbc.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Supported Database Sharding Storage
+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.
+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.
+
+## Supported version
+5.0.0-beta +
+
+## Requirement
+- Manually import MySQL Driver package mysql-connector-java-5.1.36.jar to collector-libs directory.
```
Comment (wu-sheng) [inline docs/en/Shardingjdbc.md]: `collector-libs` is not the directory name.

File: docs/en/Shardingjdbc.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Supported Database Sharding Storage
+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.
+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.
+
+## Supported version
+5.0.0-beta +
+
+## Requirement
+- Manually import MySQL Driver package mysql-connector-java-5.1.36.jar to collector-libs directory.
+- In config/application.yml, close the elasticsearch configuration and open the shardingjdbc configuration, multiple data source configurations should be separated by Half Comma.
```
Comment (wu-sheng) [inline docs/en/Shardingjdbc.md]: Considering the sharding settings have been remove from application.yml. please add the settings in here.

**Comment by chidaodezhongsheng** [comment]
@wu-sheng thanks a lot, I will resolve the conflicts and mentions.

技术部·架构部·林嘉琦
手机：18710075837
Email：linjiaqi@dangdang.com<mailto:linjiaqi@dangdang.com>
QQ：928926755

发件人: 吴晟 Wu Sheng [mailto:notifications@github.com]
发送时间: 2018年5月14日 15:45
收件人: apache/incubator-skywalking
抄送: 林嘉琦; Mention
主题: Re: [apache/incubator-skywalking] support shardingjdbc database storage feature (#1220)


@wu-sheng commented on this pull request.

@chidaodezhongsheng<https://github.com/chidaodezhongsheng> The implementation codes are under incubating. So you will take charge. I just reviewed the licenses and documents related things.

Some things need to change.

________________________________

In apm-collector/apm-collector-boot/src/main/resources/application.yml<https://github.com/apache/incubator-skywalking/pull/1220#discussion_r187851353>:

> +    contextPath: /

+storage:

+  elasticsearch:

+    clusterName: CollectorDBCluster

+    clusterTransportSniffer: true

+    clusterNodes: localhost:9300

+    indexShardsNumber: 2

+    indexReplicasNumber: 0

+    highPerformanceMode: true

+    ttl: 7

+#storage:

+#  h2:

+#    url: jdbc:h2:~/memorydb

+#    userName: sa

+#storage:

+#  shardingjdbc:

ShardingJDBC is an incubating feature, no use cases yet. And none of the contributors are committer yet. So, by that, please don't put Sharding JDBC in default application.yml. Even it is under comments.

________________________________

In apm-dist/release-docs/LICENSE<https://github.com/apache/incubator-skywalking/pull/1220#discussion_r187858208>:

> @@ -255,6 +256,7 @@ The text of each license is also included at licenses/LICENSE-[project].txt.

     Apache: commons-configuration 1.8: https://github.com/apache/commons-configuration, Apache 2.0

     Apache: commons-io 2.4: https://github.com/apache/commons-io, Apache 2.0

     Apache: tomcat 8.5.27: https://github.com/apache/tomcat/tree/trunk, Apache 2.0

+    Apache: groovy 2.4.5-indy: https://github.com/apache/groovy, Apache 2.0

Groovy has a NOTICE file. Please change NOTICE in same folder. And follow the file structure. Ref here: https://github.com/apache/groovy/blob/master/NOTICE

________________________________

In docs/README.md<https://github.com/apache/incubator-skywalking/pull/1220#discussion_r187858505>:

> -    * [Agent or collector version upgrade](en/FAQ/Upgrade.md)

-

+## Documents

+[cn doc](README_ZH.md)<https://img.shields.io/badge/document-中文-blue.svg)%5d(README_ZH.md)>

+

+  * Getting Started

+    * [Quick start](en/Quick-start.md)

+    * [Supported middlewares, frameworks and libraries](Supported-list.md)

+      * [How to disable plugins?](en/How-to-disable-plugin.md)

+  * Advanced Features

+    * [Override settings through System.properties or System.env](en/Setting-override.md)

+    * [Direct uplink and disable naming discovery](en/Direct-uplink.md)

+    * [Open TLS](en/TLS.md)

+    * [Namespace Isolation](en/Namespace.md)

+    * [Token Authentication](en/Token-auth.md)

+    * [Open Database Sharding Storage](en/Shardingjdbc.md)

This should a section in Incubating Features. named: Use Sharding JDBC as storage implementor.

________________________________

In docs/en/Shardingjdbc.md<https://github.com/apache/incubator-skywalking/pull/1220#discussion_r187858734>:

> @@ -0,0 +1,13 @@

+# Supported Database Sharding Storage

+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.

+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.

+

+## Supported version

+5.0.0-beta +

+

+## Requirement

+- Manually import MySQL Driver package mysql-connector-java-5.1.36.jar to collector-libs directory.

collector-libs is not the directory name.

________________________________

In docs/en/Shardingjdbc.md<https://github.com/apache/incubator-skywalking/pull/1220#discussion_r187858863>:

> @@ -0,0 +1,13 @@

+# Supported Database Sharding Storage

+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.

+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.

+

+## Supported version

+5.0.0-beta +

+

+## Requirement

+- Manually import MySQL Driver package mysql-connector-java-5.1.36.jar to collector-libs directory.

+- In config/application.yml, close the elasticsearch configuration and open the shardingjdbc configuration, multiple data source configurations should be separated by Half Comma.

Considering the sharding settings have been remove from application.yml. please add the settings in here.

**Comment by chidaodezhongsheng** [comment]
@wu-sheng thanks a lot, I will resolve the conflicts and mentions.

File: docs/README_ZH.md [hunk 6]
```diff
@@ -1,42 +1,46 @@
...
+    * [中间件，框架与类库支持列表](Supported-list.md)
+        * [如何关闭特定插件](cn/How-to-disable-plugin-CN.md)
+        * [可选插件](cn/Optional-plugins-CN.md)
+  * 高级特性
+    * [通过系统启动参数进行覆盖配置](cn/Setting-override-CN.md)
+    * [服务直连(Direct uplink)及禁用名称服务(naming service)](cn/Direct-uplink-CN.md)
+    * [开启TLS](cn/TLS-CN.md)
+    * [命名空间隔离](cn/Namespace-CN.md)
+    * [基于Token认证](cn/Token-auth-CN.md)
+    * [使用Shardingjdbc作为存储实现](cn/Shardingjdbc-CN.md)
```
Comment (wu-sheng) [inline docs/README_ZH.md]: In CN doc, also need an `孵化功能` section.

File: docs/cn/Shardingjdbc-CN.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# 支持数据库分片存储
+除了默认的Elasticsearch存储外，还支持数据库分片存储，允许用户将数据存储在多个数据库中。
+注意：目前仅支持MYSQL数据库的分片存储，且由于license限制，需要用户手动引入mysql驱动包。
+
+## 版本支持
+5.0.0-beta +
```
Comment (wu-sheng) [inline docs/cn/Shardingjdbc-CN.md]: beta2? beta is under release process now.

File: docs/en/Shardingjdbc.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# Supported Database Sharding Storage
+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.
+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.
+
+## Supported version
+5.0.0-beta +
```
Comment (wu-sheng) [inline docs/en/Shardingjdbc.md]: beta2? beta is under release process now.

File: docs/en/Shardingjdbc.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+Beside the default Elasticsearch storage, it also support the database sharding storage, it allow the users to store data in multiple databases.
+Note: it only supported MYSQL database sharding, and due to the license restrictions, the users need to import MYSQL Driver manually.
+
+## Supported version
+5.0.0-beta +
+
+## Requirement
+- Manually import MySQL Driver package mysql-connector-java-5.1.36.jar to collector libs directory.
+- In config/application.yml, close the elasticsearch configuration and open the shardingjdbc configuration, multiple data source configurations should be separated by Half Comma.
+``` javascript
```
Comment (wu-sheng) [inline docs/en/Shardingjdbc.md]: It is a yml, not a javascript.

File: docs/cn/Shardingjdbc-CN.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+除了默认的Elasticsearch存储外，还支持数据库分片存储，允许用户将数据存储在多个数据库中。
+注意：目前仅支持MYSQL数据库的分片存储，且由于license限制，需要用户手动引入mysql驱动包。
+
+## 版本支持
+5.0.0-beta +
+
+## 配置要求
+- 手工导入MYSQL的驱动包mysql-connector-java-5.1.36.jar到collector libs目录下。
+- config/application.yml中，关闭elasticsearch配置，打开shardingjdbc配置，多个数据源配置用半角逗号隔开。
+``` javascript
```
Comment (wu-sheng) [inline docs/cn/Shardingjdbc-CN.md]: It is a yml, not a javascript.

File: docs/README_ZH.md (hunk 6 above)
Comment (wu-sheng) [reply docs/README_ZH.md]: File name should not be shardingJDBC.  Should be: Use-ShardingJDBC-as-storage-implementor.md

**Comment by chidaodezhongsheng** [comment]
@wu-sheng I have merged the latest files from master branch and updated the markdown files.

File: apm-collector/apm-collector-storage/collector-storage-define/src/main/java/org/apache/skywalking/apm/collector/storage/dao/ui/IServiceReferenceMetricUIDAO.java [hunk 11]
```diff
@@ -1,142 +1,142 @@
...
+ * @see org.apache.skywalking.apm.collector.storage.table.service.ServiceReferenceMetricTable
+ * @see org.apache.skywalking.apm.collector.storage.StorageModule
+ */
+public interface IServiceReferenceMetricUIDAO extends DAO {
+
+    /**
+     * Returns the service reference metrics which call the given service id
+     * that collected between start time bucket and end time bucket.
+     *
+     * <p>SQL as: select BEHIND_SERVICE_ID, sum(TRANSACTION_CALLS), sum(TRANSACTION_ERROR_CALLS),
```
Comment (peng-yongsheng) [inline apm-collector/apm-collector-storage/collector-storage-define/src/main/java/org/apache/skywalking/apm/collector/storage/dao/ui/IServiceReferenceMetricUIDAO.java]: This method's definition is query front services and their metrics by passing parameter which named behind service id. The sample sql is correct.

File: apm-collector/apm-collector-storage/collector-storage-define/src/main/java/org/apache/skywalking/apm/collector/storage/dao/ui/IServiceReferenceMetricUIDAO.java [hunk 12]
```diff
@@ -1,142 +1,142 @@
...
+     * @return not nullable result list
+     */
+    List<ServiceReferenceMetric> getFrontServices(Step step, long startTimeBucket, long endTimeBucket,
+        MetricSource metricSource, int behindServiceId);
+
+    /**
+     * Returns the service reference metrics which call from the given service id
+     * that collected between start time bucket and end time bucket.
+     *
+     * <p>SQL as: select BEHIND_SERVICE_ID, sum(TRANSACTION_CALLS), sum(TRANSACTION_ERROR_CALLS),
```
Comment (peng-yongsheng) [inline apm-collector/apm-collector-storage/collector-storage-define/src/main/java/org/apache/skywalking/apm/collector/storage/dao/ui/IServiceReferenceMetricUIDAO.java]: Ditto

**Comment by peng-yongsheng** [comment]
I have added Sharding-JDBC and commons-dbcp into the maven pom file by my local project to checked the dependencies. Indeed, only added jars: groovy, sharding, dbcp.
License and notice files have been checked, correct.

**Comment by wu-sheng** [comment]
@chidaodezhongsheng @oracle219 Could you finish this pr soon?

**Comment by chidaodezhongsheng** [comment]
@wu-sheng OK, I will finish this pr now.

**Comment by wu-sheng** [comment]
@chidaodezhongsheng You are still missing @peng-yongsheng 's review.

**Pull request closed without merging**